### PR TITLE
Introduce S3 mock server to enhance unit testing

### PR DESCRIPTION
**Pull request by elega**
### What changes are proposed in this pull request?

Introduce a s3 mock server and used it in unit testing

### Why are the changes needed?

When we doing the new metadata sync redesign project, we need a reliable way to test the interactions with s3 UFS and help us deal with corner cases. 

### Does this PR introduce any user facing changes?

N/A

### CODE REVIEW

File: underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemLegacyTest.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
+/*
```
Comment (elega) [inline underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemLegacyTest.java:1]: This isn't a new file. It was renamed from S3AUnderFileSystemTest

File: core/server/master/pom.xml [hunk 2]
```diff
@@ -117,6 +117,11 @@
       <artifactId>guava-testlib</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
Comment (elega) [inline core/server/master/pom.xml:120]: mit license

File: core/server/master/pom.xml [hunk 3]
```diff
@@ -136,6 +141,12 @@
       <groupId>org.apache.httpcomponents</groupId>
       <artifactId>httpclient</artifactId>
     </dependency>
+      <dependency>
```
Comment (Jackson-Wang-7) [inline core/server/master/pom.xml:144]: It looks like the indentation here is not aligned with the previous line.

File: tests/src/test/java/alluxio/client/fs/FileSystemS3UfsIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+import java.nio.charset.StandardCharsets;
+
+public class FileSystemS3UfsIntegrationTest extends BaseIntegrationTest {
+  private static final String TEST_CONTENT = "TestContents";
+  private static final String TEST_FILE = "test_file";
+  private static final int USER_QUOTA_UNIT_BYTES = 1000;
+  @Rule
+  public LocalAlluxioClusterResource mLocalAlluxioClusterResource =
+      new LocalAlluxioClusterResource.Builder()
+          .setProperty(PropertyKey.USER_FILE_BUFFER_BYTES, USER_QUOTA_UNIT_BYTES)
```
Comment (Jackson-Wang-7) [inline tests/src/test/java/alluxio/client/fs/FileSystemS3UfsIntegrationTest.java:51]: why do not set ufs property here?

File: tests/src/test/java/alluxio/client/fs/FileSystemS3UfsIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+  @Test
+  public void basicMetadataSync() throws IOException, AlluxioException {
+    mS3Client.putObject(TEST_BUCKET, TEST_FILE, TEST_CONTENT);
+    FileInStream fis = mFileSystem.openFile(new AlluxioURI("/" + TEST_FILE));
+    assertEquals(TEST_CONTENT, IOUtils.toString(fis, StandardCharsets.UTF_8));
+  }
+
+  @Test
+  public void basicWriteThrough() throws IOException, AlluxioException {
```
Comment (Jackson-Wang-7) [inline tests/src/test/java/alluxio/client/fs/FileSystemS3UfsIntegrationTest.java:105]: what's the essential difference between this and FileSystemMasterS3UfsTest? why we need this two class to test the sync with ufs?

File: underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemMockServerTest.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+       d1/d2/f1
+       d2/
+       d2/d1/
+       d2/d1/f1
+       d3/
+       f1
+       f2
+     */
+    assertNotNull(ufsStatuses);
+    assertEquals(12, ufsStatuses.length);
```
Comment (Jackson-Wang-7) [inline underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemMockServerTest.java:133]: Do we care about the content and order of files?

**Review by Jackson-Wang-7 — COMMENTED** [review]
Good job! I have left some comments, please check, thanks.

File: minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java [hunk 7]
```diff
@@ -224,7 +224,11 @@ protected void setupTest() throws IOException {
     String underfsAddress = Configuration.getString(PropertyKey.MASTER_MOUNT_TABLE_ROOT_UFS);
 
     // Deletes the ufs dir for this test from to avoid permission problems
-    UnderFileSystemUtils.deleteDirIfExists(ufs, underfsAddress);
+    // Do not delete the ufs root if the ufs is an object storage.
+    // In test environment, this means s3 mock is used.
+    if (!ufs.isObjectStorage()) {
+      UnderFileSystemUtils.deleteDirIfExists(ufs, underfsAddress);
```
[the remark this answers is not in the document]
Comment (elega) [reply minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java:230]: you can delete the memfs root (essentially a folder) but you cannot delete the s3 root, essentially a pseudo directory.

File: core/server/master/pom.xml (hunk 3 above)
Comment (elega) [reply core/server/master/pom.xml:144]: good catch. fixed.

File: tests/src/test/java/alluxio/client/fs/FileSystemS3UfsIntegrationTest.java (hunk 5 above)
Comment (elega) [reply tests/src/test/java/alluxio/client/fs/FileSystemS3UfsIntegrationTest.java:105]: FileSystemMasterS3UfsTest -> create a file system master with some mock components. you can access to the metadata of files only, but aren't able to read or write a file. -> more a unit test
FileSystemS3UfsintegrationTest -> create an alluxio clsuter that runs on local and do integration tests. you can do whatever you want to do as if this is a real alluxio cluster (it's real technically) -> an actual integration tests

File: underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemMockServerTest.java (hunk 6 above)
Comment (elega) [reply underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemMockServerTest.java:133]: we do. the goal of this PR is to do some scaffolding and verify the feasibility of the s3 mock server, instead of writing real unit tests to test cover the existing code base. We will refine and add more tests in the future.

File: tests/src/test/java/alluxio/client/fs/FileSystemS3UfsIntegrationTest.java (hunk 4 above)
Comment (elega) [reply tests/src/test/java/alluxio/client/fs/FileSystemS3UfsIntegrationTest.java:51]: good catch

File: tests/src/test/java/alluxio/client/fs/FileSystemS3UfsIntegrationTest.java (hunk 5 above)
Comment (Jackson-Wang-7) [reply tests/src/test/java/alluxio/client/fs/FileSystemS3UfsIntegrationTest.java:106]: ok, got it

File: underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemMockServerTest.java (hunk 6 above)
Comment (Jackson-Wang-7) [reply underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemMockServerTest.java:134]: ok, no problem. good for now.

**Review by Jackson-Wang-7 — APPROVED** [review]
LGTM

File: core/server/master/pom.xml [hunk 8]
```diff
@@ -136,6 +141,12 @@
       <groupId>org.apache.httpcomponents</groupId>
       <artifactId>httpclient</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.alluxio</groupId>
+      <artifactId>alluxio-underfs-s3a</artifactId>
+      <version>2.10.0-SNAPSHOT</version>
```
Comment (jiacheliu3) [inline core/server/master/pom.xml:147]: ${project.version}?

File: core/server/master/pom.xml [hunk 9]
```diff
@@ -136,6 +141,12 @@
...
+      <scope>compile</scope>
```
Comment (jiacheliu3) [inline core/server/master/pom.xml:148]: compile only?

**Review by jiacheliu3 — APPROVED** [review]
LGTM with some minor comments, thanks!

File: core/server/master/pom.xml (hunk 8 above)
Comment (elega) [reply core/server/master/pom.xml:147]: removed this one. this was added by mistake

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please
